File: prob 49.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x

#         self.next = None
#brute force -- get the number of nodes get the nth node and the replace the nth node
# we have 2 pointers - slow and fast and move the fast pointer to right by n + 1 elements  and this way we can access the nth element from last with slow pointer
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        #edge case 1
        if not(head):
            return
        #edge case 2 - where the head.next element is not dummy
        dummy = ListNode(0)
        dummy.next = head
        slow = dummy
        fast = dummy
        count = 0
        while count <= n:
            fast = fast.next
            count += 1
        while fast:
            slow=slow.next
            fast = fast.next
        slow.next = slow.next.next
        return dummy.next

File: prob49.java
// S30 Big N Problem #49 {Medium}
// 19. Remove Nth Node From End of List
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy=new ListNode(0);
        dummy.next=head;
        
        ListNode slow=dummy;
        ListNode fast=dummy;
        
        for(int i=0;i<=n;i++){
            fast=fast.next;
        }
        
        while(fast !=null){
            fast=fast.next;
            slow=slow.next;
        }
        
        slow.next=slow.next.next;
        return dummy.next;
    }
}